feat(api, robot-server, shared-data): add FlexStacker module to the hardware controller and robot server.

vegano1 opened the pull request:
# Overview

Adds the Flex Stacker module to the hardware controller and some other gcodes to the driver.

Closes: [EXEC-981](https://opentrons.atlassian.net/browse/EXEC-981)

## Test Plan and Hands on Testing

- [x] Connect a flex stacker to the flex and make sure the module shows up in the `GET /modules` endpoint
- [x] Connect multiple flex stackers to a flex and make sure they both show up
- [x] Make sure you can move each axis a fixed distance with `move_axis` command.
- [x] Make sure you can home each axis, except for L, with the `home_axis` command.
- [x] Make sure you can update the flex stacker using the `POST /modules/update` endpoint


## Changelog

- Add FlexStacker module to the hardware controller
- Expose FlexStacker module via the robot servers `/modules` endpoint.
- add enable_motors to the driver and module (M17)
- add set_run_current (M906) and set_ihold_current (M907) to driver
- add get_motion_params (M120) to driver
- add current to MoveParams to simplify the interface
- add STACKER_MOTION_CONFIG to keep track of default move parameters used for homing or moving an axis.
- move_in_mm and move_to_limit_switch now return MoveResult
- added move_axis and home_axis methods to the FlexStacker module
- added close_latch and open_latch methods to the FlexStacker module

## Risk assessment

Low, unreleased

[EXEC-981]: https://opentrons.atlassian.net/browse/EXEC-981?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

vegano1 committed:
save

vegano1 committed:
feat(api,robot-server,shared-data): add FlexStacker module to the hardware controller and robot server.

vegano1 committed:
add poller status getters

vegano1 committed:
Merge branch 'edge' into flex-stacker-add-module-skeleton

vegano1 committed:
Update ot3controller.py

revert debug changes

ahiuchingau commented on api/src/opentrons/drivers/flex_stacker/types.py:
```diff
@@ -90,11 +98,11 @@ def distance(self, distance: float) -> float:
-    ZE: bool
-    ZR: bool
-    LR: bool
+    XE: bool = False
```
The GCode that we use to get the limit switch statuses always return results for all axes. What are the reasoning behind having setting default values?

vegano1 replied:
makes it easier to build the obj for testing.

sfoster1 replied:
the data constraints should match real life, IMO, and not have defaults. if it's annoying to build for tests, we can make a helper function in the test file that does it.

ahiuchingau commented on api/src/opentrons/drivers/flex_stacker/types.py:
```diff
@@ -128,6 +136,58 @@ def get(self, direction: Direction) -> bool:
+        return PlatformState.UNKNOWN
+
+
+class StackerAxisState(Enum):
```
Nice - I was actually planning on this exact logic of this in FW. So we can send a GCode querying the axis state, and it will return one of these strings.

vegano1 committed:
fixes to robot-server and api

vegano1 committed:
add get motion params (M120) and enable motors (M17) to FlexStackerDriver

vegano1 committed:
- add enable_motors to driver and module
- add set_run_current and set_ihold_current to driver
- add get_motion_params to driver
- add current to MoveParams to simplify interface
- add STACKER_MOTION_CONFIG to keep track of default move parameters used for homing or moving an axis.
- move_in_mm and move_to_limit_switch now return MoveResult
- added move_axis and home_axis methods to the FlexStacker module
- added close_latch and open_latch methods to the FlexStacker module

vegano1 committed:
Merge branch 'edge' into flex-stacker-add-module-skeleton

SyntaxColoring reviewed (approved):
`robot-server` and `api/src/opentrons/protocol_engine` changes LGTM. Thank you!

sfoster1 commented on api/src/opentrons/drivers/flex_stacker/abstract.py:
```diff
@@ -87,3 +105,7 @@ async def set_led(
         """Set LED color of status bar."""
         ...
+
+    async def enter_programming_mode(self) -> None:
```
nit: can we call this `prep_for_update()` or something that's more like what other modules use?

vegano1 replied:
`enter_programming_mode` is the lower-level driver function called by the module's higher-level `prep_for_update` function, which adheres to the rest of the modules.

sfoster1 reviewed (approved):
Looks good but let's not do those defaults for testing - make a helper function in the test file if needed instead.

vegano1 committed:
- remove defaults from LimitSwitchStaus and PlatformStatus
- added HopperDoorState and LatchState to FlexStacker module
- replaced hopperDoorClosed with hopperDoorState in live data
- replaced axisStateL with LatchState in live data
- use move_axis instead of home_axis for open_latch and close_latch
- set find_dfu_device expected_device_count to 3
- cleanup

caila-marashaj commented on api/src/opentrons/drivers/flex_stacker/types.py:
```diff
@@ -128,11 +140,33 @@ def get(self, direction: Direction) -> bool:
-    max_speed: float | None = None
-    acceleration: float | None = None
-    max_speed_discont: float | None = None
+    axis: Optional[StackerAxis] = None
```
When wouldn't you specify an Axis?

vegano1 replied:
probably never, ill change it in a follow-up pr I have.

vegano1 committed:
fix unit tests

caila-marashaj reviewed (approved):
Nice! I think this is gonna make the module code a little easier to structure going forward
